### PR TITLE
#130 - Add Support for Multiple User Emails

### PR DESCRIPTION
**Pull request by gamesover**
This PR introduces the ability for users to associate multiple email addresses with their account. It addresses issue #130, which highlighted the need for this feature to prevent duplicate accounts and improve user experience, particularly for users who sign up with different emails (e.g., work and personal).

**Key Changes:**

*   **Model Changes:**
    *   Created a new `Email` model to store multiple email addresses associated with a user (`belongs_to :user`).
    *   Modified the `User` model to `has_many :emails` (must be declared before devise `multi_email_authenticatable`).

*   **Devise Integration:**
    *   Added the `devise-multi_email` gem to leverage its features for managing multiple emails within Devise.

*   **Controller Changes:**
    *   Created a new `My::EmailsController` to handle:
        *   Adding new email addresses.
        *   Setting a primary email.
        *   Deleting email addresses.
        *   Sending confirmation emails upon adding a new email.
    *   Updated `ReactivationsController` and `MailingList::Sync` to use the user's primary email from the `Email` model.
*   **Mailer Changes:**
    *   Added a `trigger_after_confirmation` callback to the `Email` model to handle email updates after confirmation.
*   **View Changes:**
    *   Added views for managing multiple emails in the `my/emails` directory.
    *   Updated the `my/details/show` view to display the user's associated email addresses and their confirmation status.
    *   Added links for setting a primary email and deleting email addresses (for confirmed emails only).
*   **Migration:**
    *   Created a migration to add the `emails` table.
    *   Created a migration to temporarily remove the `null: false` constraint from the `users.email` column.
*   **Testing:**
    *   Added tests for the `Email` model, `My::EmailsController`, and other relevant components to ensure proper functionality.
    *   Updated existing tests to work with the new multi-email setup.

**Migration Strategy for Existing Users:**

1. The `users.email` column will temporarily allow null values.
2. run `User.without_emails.find_each(&:update_emails)` in rails console
3. After the data migration, the `users.email` column will be removed or repurposed.

**Further Considerations (Out of Scope for this PR):**

*   Implementing a "one-click" email association feature for a smoother user experience.
*   Updating other parts of the application that might rely on `User.email` (e.g., notifications, other mailers).
*   Adding a mechanism for users to choose which email to use for specific types of notifications.
*   Finalizing the handling of the original `email` column in the `User` model.

## UI changes

add new email
image

list emails
image

delete email
image

### CODE REVIEW

File: app/packs/entrypoints/application.js [hunk 1]
```diff
@@ -10,7 +10,8 @@
 import 'core-js/stable'
 import 'regenerator-runtime/runtime'
 
-require("@rails/ujs").start()
+import Rails from "@rails/ujs"
+Rails.start()
```
Comment (gamesover) [inline app/packs/entrypoints/application.js:14]: in my local and production, I can see below error in chrome console

image

using import to fix the error

File: app/controllers/my/emails_controller.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  end
+
+  def set_primary
+    email = Email.find(params[:id])
+    ActiveRecord::Base.transaction do
+      current_user.emails.where(primary: true).find_each { |email| email.update!(primary: false) }
+      email.update!(primary: true)
+    end
+
+    flash[:notice] = "Your primary email has been ukpdated to #{email.email}"
```
Comment (leesheppard) [inline app/controllers/my/emails_controller.rb:22]: change 'ukpdated' to updated

File: app/controllers/my/emails_controller.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    redirect_to my_details_path
+  end
+
+  def destroy
+    email = Email.find(params[:id])
+    if email.destroy
+      flash[:notice] = "Your email #{email.email} have been deleted."
```
Comment (leesheppard) [inline app/controllers/my/emails_controller.rb:29]: change **have** to **has**

File: app/controllers/my/emails_controller.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    else
+      flash[:error] = "Your email #{email.email} is failed to be deleted."
```
Comment (leesheppard) [inline app/controllers/my/emails_controller.rb:31]: Remove **is**

File: app/controllers/my/emails_controller.rb (hunk 2 above)
Comment (gamesover) [reply app/controllers/my/emails_controller.rb:22]: eagle eye :) thank you

**Comment by quintrino** [comment]
@leesheppard before you deploy this can we make sure we get a database dump as a backup?

**Comment by leesheppard** [comment]
@gamesover Error on running

```
> User.without_emails.find_each(&:update_emails)

app/lib/mailing_list/subscribe.rb:16:in `call': The CreateSend API responded with the following error - 1: Please provide a valid email address. (CreateSend::BadRequest)
        from app/lib/mailing_list/subscribe.rb:5:in `call'
        from app/models/user.rb:91:in `block in subscribe_to_lists'
        from app/lib/mailing_list.rb:17:in `each'
        from app/lib/mailing_list.rb:17:in `each'
        from app/models/user.rb:88:in `subscribe_to_lists'
        from app/models/user.rb:66:in `update_mailing_list_and_memberships'
        from app/models/email.rb:23:in `trigger_after_confirmation'
        from app/models/user.rb:58:in `update_emails'
        from (ruby-au):1:in `<top (required)>'
```

**Comment by gamesover** [comment]
An error is thrown from the following method:  
```ruby
CreateSend::Subscriber.add(
      cm_auth,
      list.api_id,
      user.email,
      user.full_name,
      nil,        # custom fields
      true,       # resubscribe
      'Unchanged' # consent_to_track
    )
```
Based on the [Campaign Monitor API documentation](https://www.campaignmonitor.com/api/v3-3/clients/#adding-a-person), this likely indicates an invalid or duplicate email(I guess, unconfirmed). However, the error message still returns `Please provide a valid email address.`  

Since these are old emails, there's no need to update the mailing list and memberships.  

Please refer to https://github.com/rubyaustralia/ruby_au/pull/267 and could you kindly run after merging it:  
```ruby
User.without_emails.find_each(&:update_emails)
```  
If needed, I’m happy to SSH into the server and debug, provided you can share the login details.

**Comment by leesheppard** [comment]
@gamesover Thank you Ken, this is all up and running. Really appreciate the work you have put in for this.

**Comment by gamesover** [comment]
> @gamesover Thank you Ken, this is all up and running. Really appreciate the work you have put in for this.

Thanks for your testing and actins.
Sorry for the trouble caused.

**Pull request merged**